tests/models/test_validator.py

```diff
@@ -0,0 +1,145 @@
+
+
+@pytest.mark.parametrize("ignore_errors", [True, False])
+@pytest.mark.parametrize("fail_on_unknown", [True, False])
+def test_models_validator_validate_with_unknown_events(
+    ignore_errors, fail_on_unknown, caplog
+):
+    """Tests given events that does not match any pydantic model,
+    the validate method should write an error message.
+    And if fail_on_unknown is set to True it should raise an UnknownEventException.
+    """
+
+    result = Validator(ModelSelector()).validate(
+        [
+            json.dumps({}),
+            json.dumps({"event_source": "browser"}),
+            json.dumps({"event_source": "browser", "event_type": None}),
+        ],
+        ignore_errors,
+        fail_on_unknown,
+    )
+    with caplog.at_level(logging.ERROR):
+        if fail_on_unknown:
+            with pytest.raises(UnknownEventException):
+                list(result)
+        else:
+            assert list(result) == []
+
+    errors = ["No matching pydantic model found for input event"] * 3
+    if fail_on_unknown:
+        errors = [errors[0]]
+    assert errors == [message for _, _, message in caplog.record_tuples]
+
+
+@pytest.mark.parametrize("ignore_errors", [True, False])
+@pytest.mark.parametrize("fail_on_unknown", [True, False])
+def test_models_validator_validate_with_invalid_page_close_browser_event(
+    ignore_errors, fail_on_unknown, caplog
+):
+    """Tests given events that match a pydantic model but the model validation fails,
+    the validate method should write an error message.
+    And if ignore_errors is set to False it should raise a BadFormatException.
+    """
+
+    result = Validator(ModelSelector()).validate(
+        [json.dumps({"event_source": "browser", "event_type": "page_close"})],
+        ignore_errors,
+        fail_on_unknown,
+    )
+    with caplog.at_level(logging.ERROR):
+        if not ignore_errors:
+            with pytest.raises(BadFormatException):
+                list(result)
+        else:
+            assert list(result) == []
+
+    errors = ["Input event is not a valid PageCloseBrowserEventModel"]
+    assert errors == [message for _, _, message in caplog.record_tuples]
+
+
+@pytest.mark.parametrize(
+    "event", [ServerEventFactory(), PageCloseBrowserEventFactory()]
+)
+@pytest.mark.parametrize("ignore_errors", [True, False])
+@pytest.mark.parametrize("fail_on_unknown", [True, False])
+def test_models_validator_validate_with_valid_events(
+    event, ignore_errors, fail_on_unknown
+):
+    """Tests given a valid event the validate method should yield it."""
+
+    event_str = event.json()
+    event_dict = json.loads(event_str)
+    validator = Validator(ModelSelector())
+    result = validator.validate([event_str], ignore_errors, fail_on_unknown)
+    assert json.loads(next(result)) == event_dict
+
+
+def test_models_validator_validate_counter(caplog):
+    """Tests given multiple events the validate method should log the total and failed events."""
```

Comment:
```suggestion
    """Given multiple events the validate method should log the total and the invalid events."""
```

Reply:
Good point) Thanks) I tried here to fit the comment in one line. Replaced with: `Tests given multiple events the validate method should log the total and invalid events.` (without the second `the`)